analyzer/runtime/evm/client.go

```diff
@@ -313,21 +313,30 @@ func EVMMaybeUnmarshalEncryptedData(data []byte, result *[]byte) (*EVMEncryptedD
 		if err := cbor.Unmarshal(*result, &callResult); err != nil {
 			return nil, fmt.Errorf("unmarshal outer call result: %w", err)
 		}
-		if callResult.IsUnknown() {
-			switch call.Format {
-			case sdkTypes.CallFormatEncryptedX25519DeoxysII:
-				var resultEnvelope sdkTypes.ResultEnvelopeX25519DeoxysII
+		switch call.Format {
+		case sdkTypes.CallFormatEncryptedX25519DeoxysII:
+			var resultEnvelope sdkTypes.ResultEnvelopeX25519DeoxysII
+			switch {
+			// The result is encrypted.
+			case callResult.IsUnknown():
 				if err := cbor.Unmarshal(callResult.Unknown, &resultEnvelope); err != nil {
-					return nil, fmt.Errorf("outer call result unmarshal unknown: %w", err)
+					return nil, fmt.Errorf("outer call result unmarshal ok: %w", err)
+				}
+			// The result can be unencrypted if the input was unencrypted.
+			case callResult.IsSuccess() && !callResult.IsUnknown():
```

Comment:
Which of `callResult.IsSuccess()` and `!callResult.IsUnknown()` indicates unencrypted input? (I'm referring to your comment.)

Can a `callResult` have `IsSuccess()` and `IsUnknown()` at the same time? If not, let's just check for `IsSuccess()` here. Edit: Urgh I see how those two methods are defined. Not nice. Worth commenting on here, I think.

Why do we handle the case of tx status=failed (as opposed to status={ok, unknown}) in such a different place, [here](https://github.com/oasisprotocol/nexus/blob/5881904c1f9a6de691dabda76152f33c0d98d7f5/analyzer/runtime/extract.go#L344-L344)? Are error messages never encrypted? Also worth commenting on here..

Reply:
`!callResult.IsUnknown()` should be covered by the above case not matching

Reply:
I need to read through this thread more tho

Reply:
ok I don't get the code comment here either.

I thought the .Unknown, .Ok, and .Failed are mutually exclusive. Unknown is allowed by the SDK, but I think I heard that Sapphire moved to making the successfulness of transactions public, meaning it chooses to use .Ok and .Failed instead of .Unknown.

Additionally, I think the implication in this PR is that .Failed is always in cleartext in Sapphire? Is that the case? Is there a systematic way to tell whether .Failed is encrypted or not?

Reply:
https://github.com/oasisprotocol/oasis-sdk/blob/runtime-sdk/v0.7.4/runtime-sdk/src/callformat.rs#L253 yes it look like failure gets passed through unencrypted under this `expose_failure` setting

Reply:
ok correction: it's not Sapphire specifically doing this, newer versions of the SDK do this for the whole EVM module https://github.com/oasisprotocol/oasis-sdk/blob/runtime-sdk/v0.7.4/runtime-sdk/modules/evm/src/lib.rs#L651

Reply:
Agree with @pro-wh

Reply:
Oof, I suppose the history is tough because we need to support pre-SDK change where errors and successes were both in `Unknown`.

Reply:
pre-change will be fine, we continue to look in Unknown if it's Unknown

Reply:
> My current understanding is that if the inner sdk callResult fails, that error is bubbled up to the outer evm callResult.Failed and so we can handle failed txs in one place at the higher level

This is a good comment, can you add it as a code comment to the `default:` branch on line 338 (which is where we handle errors)?

By `callResult.Failed`, do you mean `txr.Failed` as checked [here](https://github.com/oasisprotocol/nexus/blob/5881904c1f9a6de691dabda76152f33c0d98d7f5/analyzer/runtime/extract.go#L344-L344)?

Regardless, I'm still a little confused by the flow: If the transaction was encrypted and failed, we'll detect the failure in the `default:` branch of `case` (line 338), we'll return an error, but our caller is here and it just logs a parsing failure.
https://github.com/oasisprotocol/nexus/blob/9e33faed3fc86d90edd16892b51826a38cc569fc/analyzer/runtime/extract.go#L514-L517
Should we set `chain.runtime_transactions.success`? At the very least, we probably shouldn't be logging a parsing error.

Reply:
confirmed that when the tx fails, the parsing error isn't reached since `VisitCall` sets result to nil. I'll add a coment

Reply:
it could be unknown outer + fail inner for non-evm transactions and for old transactions